[GTK][WPE] ThreadedCompositor: Switch didRenderFrameTimer to use a RunLoopObserver

nikolaszimmermann opened the pull request:
#### d0fa6302fb1073f1a843a0d3c5bc887d8cbfc37f
<pre>
Unreviewed, reverting <a href="https://commits.webkit.org/302433@main">302433@main</a> (28fa568972a4)
</pre>

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
```diff
@@ -321,12 +320,15 @@ void ThreadedCompositor::renderLayerTree()
+    ASSERT(!m_didCompositeRunLoopObserver->isScheduled());
+    RunLoop::mainSingleton().dispatch([this, protectedThis = Ref { *this }] {
+        m_didCompositeRunLoopObserver->schedule();
+    });
```
schedule is not thread safe? if we have to dispatch a source in the main thread, then the timer was better since it has higher priority. The idea of the observer here is precisely avoid scheduling a source in the main thread run loop.

carlosgcampos replied:
I think we can update RunLoopObserver::schedule to handle the platform run loop. Then, we can schedule from the compositor thread but passing the main run loop.

nikolaszimmermann replied:
When directly scheduling from `renderLayerTree`, we end up calling `flushLayers` from the composition thread, not the main thread. That's why I used the dispatch on main loop hack.  Otherwise:

```
ASSERTION FAILED: canSafelyBeUsed()
WTF/Headers/wtf/WeakRef.h(111) : T *WTF::WeakRef<WebCore::LocalFrame>::operator->() const [T = WebCore::LocalFrame, WeakPtrImpl = WTF::DefaultWeakPtrImpl]
ASSERTION FAILED: Unsafe to ref/deref from different threads
m_isOwnedByMainThread == isMainThread()
WTF/Headers/wtf/RefCounted.h(119) : void WTF::RefCountedBase::applyRefDerefThreadingCheck() const
1   0x79a8987a375a WebCore::Page::layoutIfNeeded(WTF::OptionSet<WebCore::LayoutOptions, (WTF::ConcurrencyTag)0>)
2   0x79a8987a414b WebCore::Page::updateRendering()
3   0x79a893c53390 WebKit::WebPage::updateRendering()
4   0x79a893ca134f WebKit::LayerTreeHost::flushLayers()
5   0x79a893ca2b3f WebKit::LayerTreeHost::didComposite(unsigned int)
6   0x79a893ca8bbe WTF::Detail::CallableWrapper<WebKit::ThreadedCompositor::ThreadedCompositor(WebKit::LayerTreeHost&)::$_1, void>::call()
7   0x79a8989975d2 WTF::Detail::CallableWrapper<WebCore::RunLoopObserver::schedule(void*, WTF::OptionSet<WTF::RunLoop::Activity, (WTF::ConcurrencyTag)0>)::$_0, WTF::ActivityObserver::ContinueObservation>::call()
8   0x79a895bb6b49 WTF::RunLoop::notifyActivity(WTF::RunLoop::Activity)
9   0x79a895bb689d WTF::RunLoop::runGLibMainLoopIteration(WTF::RunLoop::MayBlock)
10  0x79a895bb7095 WTF::RunLoop::run()
11  0x79a895b37408 WTF::Thread::entryPoint(WTF::Thread::NewThreadContext*)
12  0x79a895bbe099 WTF::wtfThreadEntryPoint(void*)
13  0x79a88d759aa4 start_thread
14  0x79a88d7e6a64 __clone
```

Respecting the PlatformRunLoop is nice idea, will explore it.

nikolaszimmermann replied:
Needed to adjust things a bit for cross-thread scheduling, please have a look :-)

carlosgcampos replied:
> When directly scheduling from `renderLayerTree`, we end up calling `flushLayers` from the composition thread, not the main thread. That's why I used the dispatch on main loop hack. Otherwise:
> 
> ```
> ASSERTION FAILED: canSafelyBeUsed()
> WTF/Headers/wtf/WeakRef.h(111) : T *WTF::WeakRef<WebCore::LocalFrame>::operator->() const [T = WebCore::LocalFrame, WeakPtrImpl = WTF::DefaultWeakPtrImpl]
> ASSERTION FAILED: Unsafe to ref/deref from different threads
> m_isOwnedByMainThread == isMainThread()
> WTF/Headers/wtf/RefCounted.h(119) : void WTF::RefCountedBase::applyRefDerefThreadingCheck() const
> 1   0x79a8987a375a WebCore::Page::layoutIfNeeded(WTF::OptionSet<WebCore::LayoutOptions, (WTF::ConcurrencyTag)0>)
> 2   0x79a8987a414b WebCore::Page::updateRendering()
> 3   0x79a893c53390 WebKit::WebPage::updateRendering()
> 4   0x79a893ca134f WebKit::LayerTreeHost::flushLayers()
> 5   0x79a893ca2b3f WebKit::LayerTreeHost::didComposite(unsigned int)
> 6   0x79a893ca8bbe WTF::Detail::CallableWrapper<WebKit::ThreadedCompositor::ThreadedCompositor(WebKit::LayerTreeHost&)::$_1, void>::call()
> 7   0x79a8989975d2 WTF::Detail::CallableWrapper<WebCore::RunLoopObserver::schedule(void*, WTF::OptionSet<WTF::RunLoop::Activity, (WTF::ConcurrencyTag)0>)::$_0, WTF::ActivityObserver::ContinueObservation>::call()
> 8   0x79a895bb6b49 WTF::RunLoop::notifyActivity(WTF::RunLoop::Activity)
> 9   0x79a895bb689d WTF::RunLoop::runGLibMainLoopIteration(WTF::RunLoop::MayBlock)
> 10  0x79a895bb7095 WTF::RunLoop::run()
> 11  0x79a895b37408 WTF::Thread::entryPoint(WTF::Thread::NewThreadContext*)
> 12  0x79a895bbe099 WTF::wtfThreadEntryPoint(void*)
> 13  0x79a88d759aa4 start_thread
> 14  0x79a88d7e6a64 __clone
> ```

Of course, didComposite needs to be called in the main thread, what I suggested was to schedule from the compositor thread an observer on the main thread run loop.

> Respecting the PlatformRunLoop is nice idea, will explore it.

carlosgcampos commented on Source/WebCore/platform/glib/RunLoopObserverGLib.cpp:
```diff
@@ -29,24 +29,33 @@
-    if (m_runLoopObserver)
+    if (!runLoop)
+        runLoop = RunLoop::currentSingleton();
+    ASSERT(runLoop);
```
If the given parameter is nullptr we use currentSingleton that returns a reference, so I don't think this assert is needed and it's confusing.

nikolaszimmermann replied:
Heh, indeed. I had the assert there before adding the if (!runLoop), until the assertion fired, as I forgot about the default value being nullptr :-)

carlosgcampos commented on Source/WebCore/platform/glib/RunLoopObserverGLib.cpp:
```diff
@@ -29,24 +29,33 @@
 
-    RunLoop::currentSingleton().observeActivity(*m_runLoopObserver);
+    m_scheduledOnRunLoop = runLoop;
+    m_scheduledOnRunLoop->observeActivity(*m_runLoopObserver);
```
Since we always observe right after create and unobserve right before destroy, why don't we handle the run loop from ActivityObserver? We pass it to the constructor, so that the ActivityObserver is always attached to a run loop and it internally calls observe from constructor and unobserve from destructor.

nikolaszimmermann replied:
Will explore that.

carlosgcampos commented on Source/WTF/wtf/glib/ActivityObserver.h:
```diff
@@ -21,37 +21,52 @@
+    void attachObserverToRunLoop()
+    {
+        m_runLoop->observeActivity(*this);
+    }
```
We could make this private and call this from create() since it's the only way to create an ActivityObserver it will always be called. This way the caller doesn't need to care about attaching.

nikolaszimmermann replied:
Neat.

carlosgcampos commented on Source/WebCore/platform/glib/RunLoopObserverGLib.cpp:
```diff
@@ -29,25 +29,27 @@
+    }
 
-    m_runLoopObserver = ActivityObserver::create(static_cast<uint8_t>(m_order), activities, [this]() {
+    m_runLoopObserver = ActivityObserver::create(Ref { *runLoop.get() }, static_cast<uint8_t>(m_order), activities, [this]() {
```
Can we use releaseNonNull()?

carlosgcampos commented on Source/WTF/wtf/glib/ActivityObserver.h:
```diff
@@ -21,37 +21,54 @@
 private:
-    ActivityObserver(uint8_t order, OptionSet<RunLoop::Activity> activities, Callback&& callback)
-        : m_order(order)
+    friend Ref<ActivityObserver> create(Ref<RunLoop>&&, uint8_t, OptionSet<RunLoop::Activity>, Callback&&);
```
Why do we need this?

nikolaszimmermann replied:
Oops leftover, from when I called attachObserver.. from create(). Will remove.

carlosgcampos commented on Source/WebCore/platform/glib/RunLoopObserverGLib.cpp:
```diff
@@ -29,23 +29,28 @@
 {
     if (m_runLoopObserver) {
-        RunLoop::currentSingleton().unobserveActivity(*m_runLoopObserver);
+        m_runLoopObserver->removeObserverFromRunLoop();
```
I think we need to protect m_runLoopObserver with a lock, because we call schedule and invalidate from different threads

nikolaszimmermann replied:
Good catch, yikes.

carlosgcampos commented on Source/WTF/wtf/glib/ActivityObserver.h:
```diff
@@ -26,32 +26,45 @@
     }
 
 private:
+    Ref<RunLoop> m_runLoop;
```
Could you remind me why we don't use a thread safe weak ptr? The run loop keeps a reference to the activity observer and the activity observer to the run loop, isn't this a circular dependency? We don't need to keep the run loop alive only to unobserve the activity, no?

nikolaszimmermann replied:
Done.

carlosgcampos commented on Source/WTF/wtf/glib/ActivityObserver.h:
```diff
@@ -26,32 +26,45 @@
+        return adoptRef(*new ActivityObserver(WTFMove(runLoop), order, activities, WTFMove(callback)));
     }
 
+    void addObserverToRunLoop()
```
From a method called addObserverToRunLoop I would expect a run loop to be passed as parameter, but since the observer already knows what run loop to attach, what about using just attach/detach, or start/stop?

nikolaszimmermann replied:
Good idea, attach/detach.

carlosgcampos commented on Source/WebCore/platform/glib/RunLoopObserverGLib.cpp:
```diff
@@ -29,29 +29,33 @@
     });
 
-    RunLoop::currentSingleton().observeActivity(*m_runLoopObserver);
+    m_runLoopObserver->attachObserverToRunLoop();
```
I meant to remove the ToRunLoop() since that sounds like we are passing the run loop to attach too, but ActivityObserver already knows about the run loop. And also Observer, attachObserver sounds like attaching another thing, but the class is already the observer, so I think m_runLoopObserver->attach() or maybe better m_runLoopObserver->start() because at that point we start observing.

nikolaszimmermann replied:
Let's go for start/stop - I understand and agree with the reasoning.

carlosgcampos commented on Source/WebCore/platform/glib/RunLoopObserverGLib.cpp:
```diff
@@ -29,29 +29,33 @@
 
 bool RunLoopObserver::isScheduled() const
 {
+    Locker lock { const_cast<RunLoopObserver&>(*this).m_runLoopObserverLock };
```
We normally make the lock mutable to be able to use it in const methods

nikolaszimmermann replied:
That's better indeed.

carlosgcampos commented on Source/WebCore/platform/RunLoopObserver.h:
```diff
@@ -33,12 +33,17 @@
 using PlatformRunLoop = struct __CFRunLoop*;
+#elif USE(GLIB)
+using PlatformRunLoopObserver = ActivityObserver;
+using PlatformRunLoop = ThreadSafeWeakPtr<RunLoop>;
```
I wouldn't expose the WeakPtr detail at this point, it can be a RefPtr and then ActivityObserver creates the WeakPtr at construction.

nikolaszimmermann replied:
Ok.

carlosgcampos commented on Source/WebCore/platform/RunLoopObserver.h:
```diff
@@ -33,12 +33,17 @@
 
 #if USE(GLIB_EVENT_LOOP)
+#include <wtf/Lock.h>
+#include <wtf/WeakPtr.h>
```
We don't need this here

carlosgcampos commented on Source/WebCore/platform/glib/RunLoopObserverGLib.cpp:
```diff
@@ -29,29 +29,33 @@
 
-    m_runLoopObserver = ActivityObserver::create(static_cast<uint8_t>(m_order), activities, [this]() {
+    ThreadSafeWeakPtr weakRunLoop { runLoop ? *runLoop.get() : RunLoop::currentSingleton() };
+    m_runLoopObserver = ActivityObserver::create(WTFMove(weakRunLoop), static_cast<uint8_t>(m_order), activities, [this]() {
```
I meant ActivityObserver receives a ref, because we never pass nullptr, and the constructor creates the WeakPtr, because it's an implementation detail of ActivityObserver.

nikolaszimmermann replied:
Ok, I didn't get that.

carlosgcampos commented on Source/WTF/wtf/glib/ActivityObserver.h:
```diff
@@ -43,15 +56,17 @@ class ActivityObserver : public RefCounted<ActivityObserver> {
-    ActivityObserver(uint8_t order, OptionSet<RunLoop::Activity> activities, Callback&& callback)
-        : m_order(order)
+    ActivityObserver(ThreadSafeWeakPtr<RunLoop>&& runLoop, uint8_t order, OptionSet<RunLoop::Activity> activities, Callback&& callback)
+        : m_runLoop(WTFMove(runLoop))
```
Here is where we should create the WeakPtr

nikolaszimmermann committed:
[GTK][WPE] ThreadedCompositor: Switch didRenderFrameTimer to use a RunLoopObserver
https://bugs.webkit.org/show_bug.cgi?id=301451

Reviewed by Carlos Garcia Campos.

Enhance GLib specific RunLoopObserver infrastructure to support cross-thread
scheduling, enabling ThreadedCompositor to replace its timer-based composition
notification with a deterministic RunLoopObserver approach.

- ActivityObserver is now ThreadSafeRefCounted for cross-thread safety
- ActivityObserver holds a ThreadSafeWeakPtr to the RunLoop it got attached to
- RunLoopObserver supports scheduling on a specific RunLoop (not just current)
- RunLoopObserver handles locking of the ActivityObserver to support
  cross-thread usage scenarios (schedule()/invalidate() on different threads)
- Rename ActivityObserver::ContinueObservation::Yes|No -> NotifyResult::Continue|Stop
  and handle an additional case: 'Destroyed', when the AcvitiyObserver is no longer
  scheduled, then we shouldn't fire a notification, nor try to invalidate.
- ActivityObserver::stops() now reset the callback to nullptr. If we get
  another notify()-request and detect that the AO was already stop()'ed
  then return 'Destroyed' and ignore the request.
- Add threading related API tests for RunLoopObserver.

Tests: Tools/TestWebKitAPI/Tests/WTF/glib/ActivityObserver.cpp (adapted)
       Tools/TestWebKitAPI/Tests/WebCore/glib/RunLoopObserver.cpp (8 new tests)

* Source/WTF/wtf/RunLoop.h:
* Source/WTF/wtf/glib/ActivityObserver.h:
(WTF::ActivityObserver::create):
(WTF::ActivityObserver::~ActivityObserver):
(WTF::ActivityObserver::start):
(WTF::ActivityObserver::stop):
(WTF::ActivityObserver::notify const):
(WTF::ActivityObserver::ActivityObserver):
* Source/WTF/wtf/glib/RunLoopGLib.cpp:
(WTF::RunLoop::notifyActivity):
* Source/WebCore/platform/RunLoopObserver.h:
* Source/WebCore/platform/glib/RunLoopObserverGLib.cpp:
(WebCore::RunLoopObserver::schedule):
(WebCore::RunLoopObserver::invalidate):
(WebCore::RunLoopObserver::isScheduled const):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::m_compositingRunLoop):
(WebKit::ThreadedCompositor::invalidate):
(WebKit::ThreadedCompositor::renderLayerTree):
(WebKit::ThreadedCompositor::didCompositeRunLoopObserverFired):
(WebKit::m_didRenderFrameTimer): Deleted.
(WebKit::ThreadedCompositor::didRenderFrameTimerFired): Deleted.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Tools/TestWebKitAPI/Tests/WTF/glib/ActivityObserver.cpp:
(TestWebKitAPI::TEST(WTF_ActivityObserver, Create)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, MatchingActivity)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, NonMatchingActivity)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, MultipleActivities)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, Ordering)):
(TestWebKitAPI::TEST(WTF_ActivityObserver, SameOrder)):
* Tools/TestWebKitAPI/Tests/WebCore/glib/RunLoopObserver.cpp:
(TestWebKitAPI::SecondaryRunLoopThread::SecondaryRunLoopThread):
(TestWebKitAPI::SecondaryRunLoopThread::~SecondaryRunLoopThread):
(TestWebKitAPI::SecondaryRunLoopThread::runLoop const):
(TestWebKitAPI::SecondaryRunLoopThread::threadId const):
(TestWebKitAPI::SecondaryRunLoopThread::dispatch):
(TestWebKitAPI::TEST(RunLoopObserver, CrossThread_ScheduleOnMainLoopFromSecondaryThread)):
 -> Schedule observer from secondary thread to main loop, verify it fires on main thread.

(TestWebKitAPI::TEST(RunLoopObserver, CrossThread_ScheduleOnSecondaryLoopFromSecondaryThread)):
 -> Schedule observer from secondary thread to its own loop, verify execution on correct thread.

(TestWebKitAPI::TEST(RunLoopObserver, CrossThread_ScheduleOnSecondaryLoopFromMainThread)):
 -> Schedule observer from main thread to secondary loop.

(TestWebKitAPI::TEST(RunLoopObserver, CrossThread_TwoObserversBothOnMainLoop)):
 -> Two observers scheduled from different threads to main loop, verify both fire on main thread.

(TestWebKitAPI::TEST(RunLoopObserver, CrossThread_TwoObserversOnDifferentLoops)):
 -> Two observers on separate loops (main + secondary), verify concurrent execution on respective threads.

(TestWebKitAPI::TEST(RunLoopObserver, CrossThread_InvalidateFromDifferentThread)):
 -> Invalidate observer from different thread, verify thread-safe invalidation stops observer.

(TestWebKitAPI::TEST(RunLoopObserver, CrossThread_OneShotObserverOnSecondaryLoop)):
 -> One-shot observer on secondary loop fires exactly once across multiple iterations.

(TestWebKitAPI::TEST(RunLoopObserver, CrossThread_MultipleScheduleAttempts)):
 -> Attempt to reschedule already-scheduled observer from different thread, verify first schedule wins.

Canonical link: https://commits.webkit.org/302433@main
